src/main/scala/akka/persistence/cassandra/snapshot/CassandraSnapshotStore.scala

```diff
@@ -166,44 +165,20 @@ class CassandraSnapshotStore(cfg: Config) extends SnapshotStore with CassandraSt
   }
 
   private def metadata(prepStmt: PreparedStatement, persistenceId: String, criteria: SnapshotSelectionCriteria): Future[Vector[SnapshotMetadata]] = {
-    // TODO the RowIterator is using some blocking, would benefit from a rewrite
-    Future {
-      new RowIterator(prepStmt, persistenceId, criteria.maxSequenceNr).map { row =>
-        SnapshotMetadata(row.getString("persistence_id"), row.getLong("sequence_nr"), row.getLong("timestamp"))
-      }.dropWhile(_.timestamp > criteria.maxTimestamp).toVector
-    }(blockingDispatcher)
-  }
-
-  private class RowIterator(prepStmt: PreparedStatement, persistenceId: String, maxSequenceNr: Long) extends Iterator[Row] {
-    var currentSequenceNr = maxSequenceNr
-    var rowCount = 0
-
-    // we know that the session is initialized, since we got the prepStmt
-    val ses = Await.result(session.underlying(), config.blockingTimeout)
-
-    // FIXME more blocking
-    var iter = newIter()
-    def newIter() = ses.execute(prepStmt.bind(persistenceId, currentSequenceNr: JLong)).iterator
-
-    @annotation.tailrec
-    final def hasNext: Boolean =
-      if (iter.hasNext)
-        true
-      else if (rowCount < maxMetadataResultSize)
-        false
-      else {
-        rowCount = 0
-        currentSequenceNr -= 1
-        iter = newIter()
-        hasNext
+    def iterate(nr: Long, buf: ListBuffer[SnapshotMetadata]): Future[Vector[SnapshotMetadata]] = {
+      session.execute(prepStmt.bind(persistenceId, nr: JLong)) flatMap { rs =>
+        var rowCount = 0
+        var nextNr = 0l
+        for (r <- rs) {
```

Comment:
The problem is that the result set iterator is blocking if it is used like this. We must use non-blocking methods, such as `rs.isExhausted` , `rs.getAvailableWithoutFetching()`, `rs.fetchMoreResults()`.

I have a hard time wrapping my head around how to do that with plain futures. With streams it can be done like this: https://github.com/lagom/lagom/blob/master/persistence/src/main/scala/com/lightbend/lagom/internal/persistence/cassandra/CassandraSessionImpl.scala#L272

The stream can be turned into a future, e.g. https://github.com/lagom/lagom/blob/master/persistence/src/main/scala/com/lightbend/lagom/internal/persistence/cassandra/CassandraSessionImpl.scala#L242

I think we should copy that stuff from Lagom.